### ISSUE TITLE
ltsp-cloud: LTSP command failed: test MAC_ADDRESS= != MAC_ADDRESS=

### ISSUE THREAD

**Issue by kvaps** [ltsp/ltsp#144]
I use latest ltsp-cloud from proposed repository and [building image](https://github.com/kvaps/kubefarm/blob/master/build/ltsp/Dockerfile) on Ubuntu 20.04.

Sometimes LTSP clients does not boot:

```
[   36.316312] ACPI: If an ACPI driver is available for this device, you should use it instead of the native driver
[   36.438273] lpc_ich: Resource conflict(s) found affecting gpio_ich
[   36.513159] mlx4_core: Mellanox ConnectX core driver v4.0-0
[   36.580056] mlx4_core: Initializing 0000:01:00.0
[   36.580258] gpio_ich gpio_ich.2.auto: GPIO from 436 to 511
[   36.701195] nvme nvme0: pci function 0000:08:00.0
[   36.865587] nvme nvme0: 8/0/0 default/read/poll queues
[   42.768251] mlx4_core 0000:01:00.0: DMFS high rate steer mode is: disabled performance optimized steering
[   42.883193] mlx4_core 0000:01:00.0: 63.008 Gb/s available PCIe bandwidth (8 GT/s x8 link)
[   43.033053] mlx4_en: Mellanox ConnectX HCA Ethernet driver v4.0-0
Begin: Loading e[   43.106120] mlx4_en 0000:01:00.0: Activating port:1
ssential drivers[   43.183139] mlx4_en: 0000:01:00.0: Port 1: Using 8 TX rings
 ... done.
Begi[   43.264319] mlx4_en: 0000:01:00.0: Port 1: Using 8 RX rings
n: Running /scri[   43.347871] mlx4_en: 0000:01:00.0: Port 1: Initializing port
pts/init-premoun[   43.432420] mlx4_en 0000:01:00.0: Activating port:2
t ... done.
Beg[   43.433454] mlx4_core 0000:01:00.0 eno1: renamed from eth0
in: Mounting roo[   43.507465] mlx4_en: 0000:01:00.0: Port 2: Using 8 TX rings
t file system ..[   43.672576] mlx4_en: 0000:01:00.0: Port 2: Using 8 RX rings
. Begin: Running[   43.756148] mlx4_en: 0000:01:00.0: Port 2: Initializing port
 /scripts/local-top ... done.
B[   43.868068] <mlx4_ib> mlx4_ib_add: mlx4_ib: Mellanox ConnectX InfiniBand driver v4.0-0
egin: Running /s[   43.869847] mlx4_core 0000:01:00.0 eno1d1: renamed from eth0
cripts/local-pre[   43.968829] <mlx4_ib> mlx4_ib_add: counter index 2 for port 1 allocated 1
[   44.150731] <mlx4_ib> mlx4_ib_add: counter index 3 for port 2 allocated 1

done.
Begin: Running /scripts/local-bottom ... done.
Begin: Running /scripts/init-bottom ... Could not parse output of: ip -o route get 192.168.67.1
Could not parse output of: ip -o route get 192.168.67.2
LTSP command failed: test MAC_ADDRESS= != MAC_ADDRESS=
Aborting ltsp
LTSP boot error! Enable DEBUG_SHELL to troubleshoot!
[   44.807513] mlx4_en: eno1d1: Link Up
```

my `ltsp.conf` is
```
[common]
DEBUG_SHELL="1"
```

but debug shell is still not working to me

**Comment by alkisg** [comment]
I do not see the root filesystem mounted in the logs, is it mounted before the last `[   44.807513] mlx4_en: eno1d1: Link Up` message?

What command line are you passing to the client?

**Comment by kvaps** [comment]
I think it might be affected that `configure_network` is called by the ltsp initrd hook, it might not have enough time to configure the interface
https://github.com/ltsp/staging/blob/7351290ccb3006ce5ef796d179b3e529267ea0c9/ltsp/client/initrd-bottom/55-initrd-bottom.sh#L49-L52

before this check is called:
https://github.com/ltsp/staging/blob/41ab409c7e54999603301ac749943b5a9cd41f10/ltsp/common/ltsp/55-config.sh#L289

**Comment by kvaps** [comment]
> I do not see the root filesystem mounted in the logs, is it mounted before the last [ 44.807513] mlx4_en: eno1d1: Link Up message?

Not sure about this, I use [HTTP booting method](https://github.com/ltsp/ltsp/pull/90)

**Comment by kvaps** [comment]
`network_vars` is called each time before running any applet:
https://github.com/ltsp/staging/blob/151a14e6c9ad95a2e95a5f181dcfccd793b591e7/ltsp/common/ltsp/55-ltsp.sh#L67-L68

but the interesting fact that it is working in most cases, on rarely shows this error

**Comment by alkisg** [comment]
Ah OK that's in your area of the code that relates to http booting;
for normal LTSP, we want to let initramfs-tools itself to configure the network and mount the root filesystem, so that we don't have to troubleshoot such issues ourselves... :)

The race condition probably happens due to udev, you should wait until the device is up before proceeding to initrd-bottom.
In general LTSP shouldn't have "configure_networking" calls, this should be done in the mount stage, not in the bottom stage, i.e. you should include an initramfs hook, not LTSP code for that...

**Comment by kvaps** [comment]
I just wanted to point the discussion on IRC of this problem:
http://irclogs.ltsp.org/?d=2020-08-12


 
```

14:05 | <alkisg>  kvaps: this is where the initramfs-tools hacks go:  ltsp/client/initrd-bottom/initramfs-tools/ltsp-hook.conf:install_ltsp_hooks()  {
14:06 | <alkisg>  You can easily add a configure_networking call to "mount" from there
14:07 | <alkisg>  (or better yet, just symlink your httproot script to initramfs-tools from there)
14:10 | <alkisg>  I'll need to go, so, to sum up: I propose that you clone  /usr/share/initramfs-tools/scripts/local-top/nbd to http, you put your  code there, and you use ltsp-hook.conf to install it when you see  httproot= in the kernel cmdline
14:10 | <alkisg>  This way you mount at the "mount" stage, and everything works fine
14:11 | <alkisg>  And the code clearly shows that it's initramfs-tools specific
14:13 | <alkisg>  So your script would go to: ./ltsp/client/initrd-bottom/initramfs-tools/httproot
```

**Comment by kvaps** [comment]
fixed in [`ltsp-cloud v20.08`](https://github.com/ltsp/staging/commits/v20.08)
